### ISSUE TITLE
Not reusing existing package despite matching spec

### ISSUE THREAD

**Issue by jjwilke** [spack/spack#11113]
I have a package I am trying to build that specs as:
````
>spack spec sparc@ifpack2-base +waterman build_type=Release %gcc@7.2.0kokkos
````
which prints
````
sparc@ifpack2-base%gcc@7.2.0kokkos build_type=Release patches=4c3c98ccfe4d23b608bba7fa05345fa27b3129a2676f9d6064f2389c256d77c5,83de0aadec643fa66dd897a17e68a7ab4d9b56da737a45ea12e4a2007dc41228 ~static+waterman arch=linux-rhel7-ppc64le 
    ^cgns@3.4.0%gcc@7.2.0kokkos build_type=RelWithDebInfo ~fortran+hdf5~int64+mpi+scoping arch=linux-rhel7-ppc64le 
    ^cmake@3.14.0%gcc@7.2.0kokkos~doc+ncurses+openssl~ownlibs~qt arch=linux-rhel7-ppc64le 
    ^hdf5@1.10.5%gcc@7.2.0kokkos~cxx~debug~fortran~hl+mpi+pic+shared~szip~threadsafe arch=linux-rhel7-ppc64le 
    ^netcdf@4.6.3%gcc@7.2.0kokkos~dap~hdf4 maxdims=1024 maxvars=8192 +mpi~parallel-netcdf+pic+shared arch=linux-rhel7-ppc64le 
    ^openblas@0.3.5%gcc@7.2.0kokkos cpu_target=auto ~ilp64+pic+shared threads=none ~virtual_machine arch=linux-rhel7-ppc64le 
    ^openmpi@3.1.3%gcc@7.2.0kokkos~cuda+cxx_exceptions fabrics=verbs ~java~legacylaunchers~memchecker~pmi schedulers=auto ~sqlite3~thread_multiple+vt arch=linux-rhel7-ppc64le 
    ^trilinos@ifpack2-base%gcc@7.2.0kokkos build_type=Release +sparc+static+waterman arch=linux-rhel7-ppc64le 
        ^boost@1.69.0%gcc@7.2.0kokkos+atomic+chrono~clanglibcpp~context~coroutine cxxstd=98 +date_time~debug+exception~fiber+filesystem+graph~icu+iostreams+locale+log+math~mpi+multithreaded~numpy patches=2ab6c72d03dec6a4ae20220a9dfd5c8c572c5294252155b85c6874d97c323199,3a83d907043708218325c35ffc318fd6d6cfd78ba89a78f2c70013c72603e5b8,607b0772dec1287c9084ae3b36ee32bff945a2fe5e608823ed47a1ea765c84cd ~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave arch=linux-rhel7-ppc64le 
        ^parmetis@4.0.3%gcc@7.2.0kokkos build_type=RelWithDebInfo ~gdb patches=4f892531eb0a807eb1b82e683a416d3e35154a455274cf9b162fb02054d11a5b,50ed2081bc939269689789942067c58b3e522c269269a430d5d34c00edbc5870,704b84f7c7444d4372cb59cca6e1209df4ef3b033bc4ee3cf50f369bce972a9d +shared arch=linux-rhel7-ppc64le 
        ^pnetcdf@system-build%gcc@7.2.0kokkos build_type=RelWithDebInfo arch=linux-rhel7-ppc64le 
        ^superlu-dist@6.1.0%gcc@7.2.0kokkos build_type=RelWithDebInfo ~int64+shared arch=linux-rhel7-ppc64le 
````

I can list my trilinos installs:
````
>spack find -v trilinos
````
which prints
````
==> 4 installed packages
-- linux-rhel7-ppc64le / gcc@7.2.0kokkos ------------------------
trilinos@cuda-mpi-apr3 build_type=Release +sparc+static+waterman
trilinos@getnorm-mar25 build_type=Release +sparc+static+waterman
trilinos@ifpack2-base build_type=Release +sparc+static+waterman
trilinos@pinned-mem-apr3 build_type=Release +sparc+static+waterman
````

When I try to install using the same spec:
````
>spack install sparc@ifpack2-base +waterman build_type=Release %gcc@7.2.0kokkos
````
trilinos gets rebuilt and installed. Am I missing something on the specs potentially not matching? Any help would be greatly appreciated.

**Comment by jjwilke** [comment]
UPDATED using --very-long. The spec generates the hash
````
m42r756q2uci4mnvmqf5ppxt4z3spibz      
^trilinos@ifpack2-base%gcc@7.2.0kokkos build_type=Release +sparc+static+waterman arch=linux-rhel7-ppc64le 
````
The existing hash for the installed trilinos is:
````
-- linux-rhel7-ppc64le / gcc@7.2.0kokkos ------------------------
ce25t3myupvnr7fv3vulm7kaci37k5wl trilinos@ifpack2-base build_type=Release +sparc+static+waterman
````
I assume this is the cause? Why would these hashes end up different with exactly the same spec.

**Comment by jjwilke** [comment]
UPDATED:
Somehow one of the hashes of the dependencies of trilinos was different. I now have for output of spec/find:
````
7mye3l4fbjc3h4vjcilrxfl6zfp7dnof      ^trilinos@ifpack2-base%gcc@7.2.0kokkos build_type=Release +sparc+static+waterman arch=linux-rhel7-ppc64le
````
and
````
-- linux-rhel7-ppc64le / gcc@7.2.0kokkos ------------------------
dgjgmxygjq327hgrv3xr2ectv72qs3vp trilinos@cuda-mpi-apr3  
7mye3l4fbjc3h4vjcilrxfl6zfp7dnof trilinos@ifpack2-base
ynlqxxwsw4xknmoyeeq6c3edhururobp trilinos@getnorm-mar25  b4m3z5k4kwovm4igizbhqw5g5nyl7ku7 trilinos@pinned-mem-apr3
````
Spack is still trying to rebuild trilinos instead of using an exact hash match.

**Comment by becker33** [comment]
Try comparing the ouptput from `spack spec <your abstract spec>` to `spack find -ldvf trilinos`

My guess is there's some variant or dependency you're setting differently than the existing build.

**Comment by jjwilke** [comment]
Yeah, we can go ahead and close this. I'm pretty sure I foo-barred a single dependency.
This might be better as a feature request.

`spack tell-me-why-no-match`

`spack no-i-promise-this-is-the-same`

Is there already some workflow for this and I just need to RTDM? I don't have a very complex environment (yet). But this would be pretty upsetting if I have a big chain of dependencies and suddenly had to rebuild EVERYTHING to fix one bad hash. I guess the hashes are linked to the install paths and there would be all sorts RPATH, etc, issues around trying to patch one hash the got messed up.

**Comment by becker33** [comment]
We have talked about using the binary packaging capability (rewriting RPATHs) to have a command like `spack replace spec dependency_spec` which would reinstall `spec` from binary but rewriting its RPATHs to use `dependency_spec` instead of whatever spec it used for `dependency` originally. But we don't have any concrete plans to pursue that on any particular timeframe.

A spec diff tool is also a good idea, but we haven't thought about that in any serious way so far as I'm aware.

Closing this as resolved, feel free to add feature requests for either feature as I don't think they're already up.
